src/main/scala/is/hail/io/vcf/LoadVCF.scala

```diff
@@ -109,27 +109,27 @@ object LoadVCF {
         (line.getType == VCFHeaderLineType.Flag && line.getCount == 0)))
       Field(id, baseType, i, attrs)
     else
-      Field(id, TArray(baseType), i, attrs)
+      Field(id, TArray(if (arrayElementsRequired) !baseType else baseType), i, attrs)
   }
 
   def headerSignature[T <: VCFCompoundHeaderLine](lines: java.util.Collection[T],
-    callFields: Set[String] = Set.empty[String]): TStruct = {
+    callFields: Set[String] = Set.empty[String], arrayElementsRequired: Boolean = true): TStruct = {
     TStruct(lines
       .zipWithIndex
-      .map { case (line, i) => headerField(line, i, callFields) }
+      .map { case (line, i) => headerField(line, i, callFields, arrayElementsRequired = arrayElementsRequired) }
       .toArray)
   }
 
   def formatHeaderSignature[T <: VCFCompoundHeaderLine](lines: java.util.Collection[T],
-    callFields: Set[String] = Set.empty[String]): (TStruct, Int) = {
+    callFields: Set[String] = Set.empty[String], arrayElementsRequired: Boolean = true): (TStruct, Int) = {
     val canonicalFields = Array(
       "GT" -> TCall(),
-      "AD" -> TArray(TInt32()),
+      "AD" -> TArray(if (arrayElementsRequired) !TInt32() else TInt32()),
```

Comment:
In this case, isn't it clearer to use the `TInt32` constructor that takes required rather than using `!`?

Reply:
you mean TInt32(arrayElementsRequired)? yeah, probably. I'll change.